### ISSUE TITLE
OpenCL unavailable if CV 4.6.0 compiled with TIMVX

### ISSUE THREAD

**Issue by simoberny** [opencv/opencv#23016]
### System Information

OpenCV version: 4.6.0
Operating System / Platform: Yocto hardknott-fsl-5.10.52_2.1.0
Compiler & compiler version: GCC 10.2.0 / CMake 3.19.5
Variscite IMX.8M Plus compute unit on custom carrier board

### Detailed description

I have compiled OpenCV using the newest version (4.6.0) to exploit the board's internal NPU with TIM-VX, and everything is working as expected. Indeed, the inference is spectacular. 

To optimize the computation time further I tried to use OpenCL on the internal GPU. 
The strange behavior is that if I compile OpenCV with TIMVX enabled, OpenCL becomes unavailable, even if the build information says otherwise. 
On the opposite, if I compile OpenCV without TIMVX the same piece of code correctly identifies OpenCL and the internal GPU. 

### Compilation
```bash
cmake -B opencv-build -D CMAKE_INSTALL_PREFIX=opencv-install -D BUILD_opencv_gapi=OFF -D BUILD_TIFF=OFF 
    -D WITH_CUDA=OFF -D WITH_OPENGL=OFF -D OPENCL_FOUND=ON -D WITH_OPENCL=ON -D WITH_IPP=OFF 
    -D WITH_TBB=ON -D BUILD_TBB=ON -D WITH_V4L=ON -D WITH_VTK=OFF -D WITH_QT=OFF -D BUILD_TESTS=OFF 
    -D BUILD_PERF_TESTS=OFF -D OPENCV_GENERATE_PKGCONFIG=ON -D INSTALL_PYTHON_EXAMPLES=OFF 
    -D INSTALL_C_EXAMPLES=OFF -D CMAKE_BUILD_TYPE=RELEASE -D WITH_FFMPEG=ON -D WITH_TIMVX=ON 
    -D TIMVX_INSTALL_DIR=/opt/timvx-build/install -D VIVANTE_SDK_DIR=/opt/aarch64_A311D_6.4.8 
    -D OPENCV_EXTRA_MODULES_PATH=/opt/opencv_contrib/modules opencv
```

### Code sample to test OpenCL
```cpp
#include <opencv4/opencv2/opencv.hpp>
#include <opencv4/opencv2/core/ocl.hpp>

using namespace cv;

void checkOpenCL() 
{
    if (!cv::ocl::haveOpenCL())
    {
        std::cout << "OpenCL is not available..." <<  std::endl;
        return;
    }

    cv::ocl::Context context;
    if (!context.create(cv::ocl::Device::TYPE_ALL))
    {
        std::cout << "Failed creating the context..." <<  std::endl;
        return;
    }

    std::cout << context.ndevices() << " CPU/GPU devices are detected." <<  std::endl;
    for (int i = 0; i < context.ndevices(); i++)
    {
        cv::ocl::Device device = context.device(i);
        std::cout << "name:              " << device.name() <<  std::endl;
        std::cout << "available:         " << device.available() <<  std::endl;
        std::cout << "imageSupport:      " << device.imageSupport() <<  std::endl;
        std::cout << "OpenCL_C_Version:  " << device.OpenCL_C_Version() <<  std::endl;
        std::cout <<  std::endl;
    }

    cv::ocl::Device(context.device(0));
}

int main(int argc, char *argv[])
{
    std::cout << cv::getBuildInformation() << std::endl;
    checkOpenCL();

    return 0;
}
```

### Result with OpenCV 4.6.0 without TIMVX
```
1 CPU devices are detected.
name:              Vivante OpenCL Device GC7000UL.6204.0000
available:         1
imageSupport:      1
OpenCL_C_Version:  OpenCL C 1.2 
```

<details>
  <summary>Build information</summary>

###
  ```
General configuration for OpenCV 4.6.0 =====================================
  Version control:               4.6.0

  Extra modules:
    Location (extra):            /opt/opencv_contrib/modules
    Version control (extra):     4.6.0-dirty

  Platform:
    Timestamp:                   2022-12-22T09:42:32Z
    Host:                        Linux 5.10.52+g0fa0413c44d5 aarch64
    CMake:                       3.19.5
    CMake generator:             Unix Makefiles
    CMake build tool:            /usr/bin/make
    Configuration:               RELEASE

  CPU/HW features:
    Baseline:                    NEON FP16

  C/C++:
    Built as dynamic libs?:      YES
    C++ standard:                11
    C++ Compiler:                /usr/bin/c++  (ver 10.2.0)
    C++ flags (Release):         -fsigned-char -W -Wall -Wreturn-type -Wnon-virtual-dtor -Waddress -Wsequence-point -Wformat -Wformat-security -Wmissing-declarations -Wundef -Winit-self -Wpointer-arith -Wshadow -Wsign-promo -Wuninitialized -Wsuggest-override -Wno-delete-non-virtual-dtor -Wno-comment -Wimplicit-fallthrough=3 -Wno-strict-overflow -fdiagnostics-show-option -pthread -fomit-frame-pointer -ffunction-sections -fdata-sections    -fvisibility=hidden -fvisibility-inlines-hidden -O3 -DNDEBUG  -DNDEBUG
    C++ flags (Debug):           -fsigned-char -W -Wall -Wreturn-type -Wnon-virtual-dtor -Waddress -Wsequence-point -Wformat -Wformat-security -Wmissing-declarations -Wundef -Winit-self -Wpointer-arith -Wshadow -Wsign-promo -Wuninitialized -Wsuggest-override -Wno-delete-non-virtual-dtor -Wno-comment -Wimplicit-fallthrough=3 -Wno-strict-overflow -fdiagnostics-show-option -pthread -fomit-frame-pointer -ffunction-sections -fdata-sections    -fvisibility=hidden -fvisibility-inlines-hidden -g  -O0 -DDEBUG -D_DEBUG
    C Compiler:                  /usr/bin/cc
    C flags (Release):           -fsigned-char -W -Wall -Wreturn-type -Waddress -Wsequence-point -Wformat -Wformat-security -Wmissing-declarations -Wmissing-prototypes -Wstrict-prototypes -Wundef -Winit-self -Wpointer-arith -Wshadow -Wuninitialized -Wno-comment -Wimplicit-fallthrough=3 -Wno-strict-overflow -fdiagnostics-show-option -pthread -fomit-frame-pointer -ffunction-sections -fdata-sections    -fvisibility=hidden -O3 -DNDEBUG  -DNDEBUG
    C flags (Debug):             -fsigned-char -W -Wall -Wreturn-type -Waddress -Wsequence-point -Wformat -Wformat-security -Wmissing-declarations -Wmissing-prototypes -Wstrict-prototypes -Wundef -Winit-self -Wpointer-arith -Wshadow -Wuninitialized -Wno-comment -Wimplicit-fallthrough=3 -Wno-strict-overflow -fdiagnostics-show-option -pthread -fomit-frame-pointer -ffunction-sections -fdata-sections    -fvisibility=hidden -g  -O0 -DDEBUG -D_DEBUG
    Linker flags (Release):      -Wl,--gc-sections -Wl,--as-needed -Wl,--no-undefined  
    Linker flags (Debug):        -Wl,--gc-sections -Wl,--as-needed -Wl,--no-undefined  
    ccache:                      YES
    Precompiled headers:         NO
    Extra dependencies:          dl m pthread rt
    3rdparty dependencies:

  OpenCV modules:
    To be built:                 aruco barcode bgsegm bioinspired calib3d ccalib core datasets dnn dnn_objdetect dnn_superres dpm face features2d flann fuzzy hfs highgui img_hash imgcodecs imgproc intensity_transform line_descriptor mcc ml objdetect optflow phase_unwrapping photo plot python3 quality rapid reg rgbd saliency shape stereo stitching structured_light superres surface_matching text tracking video videoio videostab wechat_qrcode xfeatures2d ximgproc xobjdetect xphoto
    Disabled:                    gapi world
    Disabled by dependency:      -
    Unavailable:                 alphamat cudaarithm cudabgsegm cudacodec cudafeatures2d cudafilters cudaimgproc cudalegacy cudaobjdetect cudaoptflow cudastereo cudawarping cudev cvv freetype hdf java julia matlab ovis python2 sfm ts viz
    Applications:                apps
    Documentation:               NO
    Non-free algorithms:         NO

  GUI:                           NONE
    GTK+:                        NO

  Media I/O: 
    ZLib:                        /usr/lib/libz.so (ver 1.2.11)
    JPEG:                        libjpeg-turbo (ver 2.1.2-62)
    WEBP:                        build (ver encoder: 0x020f)
    PNG:                         build (ver 1.6.37)
    TIFF:                        build (ver 42 - 4.2.0)
    JPEG 2000:                   build (ver 2.4.0)
    OpenEXR:                     build (ver 2.3.0)
    HDR:                         YES
    SUNRASTER:                   YES
    PXM:                         YES
    PFM:                         YES

  Video I/O:
    DC1394:                      NO
    FFMPEG:                      NO
      avcodec:                   NO
      avformat:                  NO
      avutil:                    NO
      swscale:                   NO
      avresample:                NO
    GStreamer:                   NO
    v4l/v4l2:                    YES (linux/videodev2.h)

  Parallel framework:            TBB (ver 2020.2 interface 11102)

  Trace:                         YES (with Intel ITT)

  Other third-party libraries:
    Lapack:                      NO
    Eigen:                       NO
    Custom HAL:                  YES (carotene (ver 0.0.1))
    Protobuf:                    build (3.19.1)

  OpenCL:                        YES (no extra features)
    Include path:                /opt/opencv/3rdparty/include/opencl/1.2
    Link libraries:              Dynamic load

  Python 3:
    Interpreter:                 /usr/bin/python3 (ver 3.9.5)
    Libraries:                   /usr/lib/libpython3.9.so (ver 3.9.5)
    numpy:                       /usr/lib/python3.9/site-packages/numpy/core/include (ver 1.20.1)
    install path:                lib/python3.9/site-packages/cv2/python-3.9

  Python (for build):            /usr/bin/python3

  Java:                          
    ant:                         NO
    JNI:                         NO
    Java wrappers:               NO
    Java tests:                  NO

  Install to:                    /opt/opencv-install
```
</details>

### Result with OpenCV 4.6.0 with TIMVX
```
OpenCL is not available...
Failed creating the context...
0 CPU devices are detected.
```

<details>
  <summary>Build information</summary>

###
  ```
General configuration for OpenCV 4.6.0 =====================================
  Version control:               4.6.0

  Extra modules:
    Location (extra):            /opt/opencv_contrib/modules
    Version control (extra):     4.6.0-dirty

  Platform:
    Timestamp:                   2022-12-22T10:16:44Z
    Host:                        Linux 5.10.52+g0fa0413c44d5 aarch64
    CMake:                       3.19.5
    CMake generator:             Unix Makefiles
    CMake build tool:            /usr/bin/make
    Configuration:               RELEASE

  CPU/HW features:
    Baseline:                    NEON FP16

  C/C++:
    Built as dynamic libs?:      YES
    C++ standard:                11
    C++ Compiler:                /usr/bin/c++  (ver 10.2.0)
    C++ flags (Release):         -fsigned-char -W -Wall -Wreturn-type -Wnon-virtual-dtor -Waddress -Wsequence-point -Wformat -Wformat-security -Wmissing-declarations -Wundef -Winit-self -Wpointer-arith -Wshadow -Wsign-promo -Wuninitialized -Wsuggest-override -Wno-delete-non-virtual-dtor -Wno-comment -Wimplicit-fallthrough=3 -Wno-strict-overflow -fdiagnostics-show-option -pthread -fomit-frame-pointer -ffunction-sections -fdata-sections    -fvisibility=hidden -fvisibility-inlines-hidden -O3 -DNDEBUG  -DNDEBUG
    C++ flags (Debug):           -fsigned-char -W -Wall -Wreturn-type -Wnon-virtual-dtor -Waddress -Wsequence-point -Wformat -Wformat-security -Wmissing-declarations -Wundef -Winit-self -Wpointer-arith -Wshadow -Wsign-promo -Wuninitialized -Wsuggest-override -Wno-delete-non-virtual-dtor -Wno-comment -Wimplicit-fallthrough=3 -Wno-strict-overflow -fdiagnostics-show-option -pthread -fomit-frame-pointer -ffunction-sections -fdata-sections    -fvisibility=hidden -fvisibility-inlines-hidden -g  -O0 -DDEBUG -D_DEBUG
    C Compiler:                  /usr/bin/cc
    C flags (Release):           -fsigned-char -W -Wall -Wreturn-type -Waddress -Wsequence-point -Wformat -Wformat-security -Wmissing-declarations -Wmissing-prototypes -Wstrict-prototypes -Wundef -Winit-self -Wpointer-arith -Wshadow -Wuninitialized -Wno-comment -Wimplicit-fallthrough=3 -Wno-strict-overflow -fdiagnostics-show-option -pthread -fomit-frame-pointer -ffunction-sections -fdata-sections    -fvisibility=hidden -O3 -DNDEBUG  -DNDEBUG
    C flags (Debug):             -fsigned-char -W -Wall -Wreturn-type -Waddress -Wsequence-point -Wformat -Wformat-security -Wmissing-declarations -Wmissing-prototypes -Wstrict-prototypes -Wundef -Winit-self -Wpointer-arith -Wshadow -Wuninitialized -Wno-comment -Wimplicit-fallthrough=3 -Wno-strict-overflow -fdiagnostics-show-option -pthread -fomit-frame-pointer -ffunction-sections -fdata-sections    -fvisibility=hidden -g  -O0 -DDEBUG -D_DEBUG
    Linker flags (Release):      -Wl,--gc-sections -Wl,--as-needed -Wl,--no-undefined  
    Linker flags (Debug):        -Wl,--gc-sections -Wl,--as-needed -Wl,--no-undefined  
    ccache:                      YES
    Precompiled headers:         NO
    Extra dependencies:          dl m pthread rt
    3rdparty dependencies:

  OpenCV modules:
    To be built:                 aruco barcode bgsegm bioinspired calib3d ccalib core datasets dnn dnn_objdetect dnn_superres dpm face features2d flann fuzzy hfs highgui img_hash imgcodecs imgproc intensity_transform line_descriptor mcc ml objdetect optflow phase_unwrapping photo plot python3 quality rapid reg rgbd saliency shape stereo stitching structured_light superres surface_matching text tracking video videoio videostab wechat_qrcode xfeatures2d ximgproc xobjdetect xphoto
    Disabled:                    gapi world
    Disabled by dependency:      -
    Unavailable:                 alphamat cudaarithm cudabgsegm cudacodec cudafeatures2d cudafilters cudaimgproc cudalegacy cudaobjdetect cudaoptflow cudastereo cudawarping cudev cvv freetype hdf java julia matlab ovis python2 sfm ts viz
    Applications:                apps
    Documentation:               NO
    Non-free algorithms:         NO

  GUI:                           NONE
    GTK+:                        NO

  Media I/O: 
    ZLib:                        /usr/lib/libz.so (ver 1.2.11)
    JPEG:                        libjpeg-turbo (ver 2.1.2-62)
    WEBP:                        build (ver encoder: 0x020f)
    PNG:                         build (ver 1.6.37)
    TIFF:                        build (ver 42 - 4.2.0)
    JPEG 2000:                   build (ver 2.4.0)
    OpenEXR:                     build (ver 2.3.0)
    HDR:                         YES
    SUNRASTER:                   YES
    PXM:                         YES
    PFM:                         YES

  Video I/O:
    DC1394:                      NO
    FFMPEG:                      NO
      avcodec:                   NO
      avformat:                  NO
      avutil:                    NO
      swscale:                   NO
      avresample:                NO
    GStreamer:                   NO
    v4l/v4l2:                    YES (linux/videodev2.h)

  Parallel framework:            TBB (ver 2020.2 interface 11102)

  Trace:                         YES (with Intel ITT)

  Other third-party libraries:
    Lapack:                      NO
    Eigen:                       NO
    Custom HAL:                  YES (carotene (ver 0.0.1))
    Protobuf:                    build (3.19.1)

  Tim-VX:                        YES
    Include path                 /opt/timvx-build/install/include
    Link libraries:              /opt/timvx-build/install/lib/libtim-vx.so
    VIVANTE SDK path             /opt/aarch64_A311D_6.4.8

  OpenCL:                        YES (no extra features)
    Include path:                /opt/opencv/3rdparty/include/opencl/1.2
    Link libraries:              Dynamic load

  Python 3:
    Interpreter:                 /usr/bin/python3 (ver 3.9.5)
    Libraries:                   /usr/lib/libpython3.9.so (ver 3.9.5)
    numpy:                       /usr/lib/python3.9/site-packages/numpy/core/include (ver 1.20.1)
    install path:                lib/python3.9/site-packages/cv2/python-3.9

  Python (for build):            /usr/bin/python3

  Java:                          
    ant:                         NO
    JNI:                         NO
    Java wrappers:               NO
    Java tests:                  NO

  Install to:                    /opt/opencv-install
```
</details>

### Issue submission checklist

- [X] I report the issue, it's not a question
- [X] I checked the problem with documentation, FAQ, open issues, forum.opencv.org, Stack Overflow, etc and have not found any solution
- [X] I updated to the latest OpenCV version and the issue is still there
- [ ] There is reproducer code and related data files (videos, images, onnx, etc)

**Comment by simoberny** [comment]
The problem persists with the 4.7.0 release. 
I did not find any additional information about that, but could it be that enabling NPU and OpenCL are mutually exclusive?

Thanks

**Comment by mshabunin** [comment]
Try to ran your application with `LD_DEBUG=libs` and `OPENCV_LOG_LEVEL=debug` environment variables set and check logs for failed _libOpenCL.so_ loads and other hints.

**Comment by simoberny** [comment]
Ok, I got the problem. 
I misunderstood the NPU wiki page by OpenCV. I installed also the A311D toolchain (even if I'm on an NXP board) so when compiling with TIMVX it was using the wrong OpenCL library from A311D folder. Compiling instead without TIMVX the correct native CL library is loaded. 

This actually means the inference was carried out by the CPU. To actually exploit the NPU I must use TFLite on this type of board. 

Thanks for the tips,
Sorry for the inconvenience
